remove cobra dependency from pkg/cmd/template package

cppforlife opened the pull request:
ytt library users do not want to bring in unnecessary dependencies (such as cobra). 

per request from #carvel slack: https://kubernetes.slack.com/archives/CH8KCCKA5/p1647894140376479

Commit:
remove cobra dependency from pkg/cmd/template package

ytt library users do not want to bring in unnecessary dependencies

Commit:
make toml an optional ytt module

so that ytt integrators do not have to take a dependency on toml

pivotaljohn commented on examples/integrating-with-ytt/internal-templating/go.mod:
```diff
@@ -2,13 +2,11 @@ module example_internal_templating
 go 1.17
 
-require github.com/k14s/ytt v0.38.0
+replace github.com/vmware-tanzu/carvel-ytt => ../../../
```
What motivates this `replace` directive?

As an example, I assume it's more useful when it's self-contained.

pivotaljohn replied:
```suggestion
replace github.com/vmware-tanzu/carvel-ytt => ../../../   // ensure example works with this copy of ytt; remove before use.
```

pivotaljohn replied:
```suggestion
replace github.com/vmware-tanzu/carvel-ytt => ../../../   // ensure example works with this copy of ytt; remove before use.
```

pivotaljohn commented on pkg/cmd/template/cmd.go:
```diff
@@ -55,26 +54,20 @@ func NewOptions() *Options {
-		RunE:    func(c *cobra.Command, args []string) error { return o.Run() },
-	}
-	cmd.Flags().BoolVar(&o.IgnoreUnknownComments, "ignore-unknown-comments", false,
+// BindFlags registers top level template flags for template command.
+func (o *Options) BindFlags(cmdFlags CmdFlags) {
```
```suggestion
// BindFlags registers template flags for template command.
func (o *Options) BindFlags(cmdFlags CmdFlags) {
```

nit: while this function technically delegates to contained structs for binding their flags, it's whole responsibility is to all flags.

pivotaljohn commented on pkg/yttlibrary/all.go:
```diff
@@ -7,18 +7,34 @@ import (
 
+var registeredExts []*starlarkstruct.Module
+
+// DoNotUseRegisterExt allows to registry ytt library extensions.
+// It's important to allow only specific module names to be registered
+// as we do not want end-users of ytt to extend @ytt library.
+// (They should define their own library for that.)
+func DoNotUseRegisterExt(mod *starlarkstruct.Module) {
```
```suggestion
// RegisterExt adds "mod" to the standard set of ytt library modules as an extension.
// An "extension" is a Starlark module that has external Go dependencies
// (as opposed to other Starlark modules in the ytt library that either have no dependencies or depend on the Go standard lib)
// This enables those using ytt as a Go module to opt-in (rather than be forced) to accept such dependencies.
// Only Carvel-maintained extensions can be registered; this reserves the `@ytt:` namespace.
// Integrators who want to write their own extensions should construct their own library.
func RegisterExt(mod *starlarkstruct.Module) {
```

Given that it contains mechanism that implements the policy, it feels a bit "overkill" to _also_ prepend the name of the function with `DoNotUse...`.
Add explicit definition of what an "extension" is; providing a well-defined concept for this overall mechanism (registering).
Also suggesting additional context for usage by maintainers.

pivotaljohn replied:
@cppforlife noted:
> the reason why ive added DoNotUse is because i want to strongly discourage anybody from using it.
> it's really a "private" method to ytt project, but go doesnt have a way to define it.

Yes. The menu of "channels" to "signal" here include:
- the method name
- Go Doc
- error messages.

I'm suggesting that the latter two are sufficient to send this signal.
Further that we're "denormalizing" the method namespace with this kind of prefix. (tho, TBH, if we would adopt something like `ModulePrivate_`, `Private` or just `_` as a prefix, I'd be _more_ okay with it — such a convention would easily apply for both integrators and contributors without (much) of a story to tell).

pivotaljohn replied:
We spoke about this.

For now, we'll adopt the changes I suggested.

At the same time, @cppforlife has some best-laid plans for providing for a proper library extensions feature for ytt. So, this will likely evolve through that work.

pivotaljohn commented on pkg/yttlibrary/all.go:
```diff
@@ -47,7 +62,17 @@ func NewAPI(replaceNodeFunc tplcore.StarlarkFunc, dataMod DataModule,
+	}
+
+	for _, ext := range registeredExts {
+		// Double check that we are not overriding predefined library
```
nice! 👍🏻

pivotaljohn commented on pkg/yttlibraryext/toml/toml.go:
```diff
@@ -14,21 +14,26 @@ import (
 )
 
+func init() {
+	yttlibrary.DoNotUseRegisterExt(tomlMod)
```
```suggestion
	yttlibrary.RegisterExt(tomlMod)
```
(see rationale at definition)

pivotaljohn reviewed (changes requested):
Minor feedback: all renaming and comment improvements.

Otherwise, looks great.

Commit:
format examples/integrating-with-ytt

cari-lynn committed:
Explicit set Go version for linting in CI

- golangci-lint-action@v2 was implicitly installing latest go. Go just
  bumped to 1.18, so import data is incompatible with 1.17.
- golangci-lint-action@v3 no long implicitly installs go.
- Bump `golangci-lint` to 1.45.2
  - this version supports _both_ go 1.17 and go 1.18; will make the
    transition to 1.18 smoother if CI is ready to go.
  - also using latest "Setup Go" action (upgrading all the things).

Signed-off-by: John Ryan <jtigger@infosysengr.com>
Signed-off-by: Cari Dean <cdean@pivotal.io>

pivotaljohn committed:
Merge branch 'develop' into ytt-no-cobra-dep
